near-merkle-proofs/src/lib.rs

```diff
@@ -0,0 +1,757 @@
+struct CachedNodes {
+    inner: HashMap<(Level, Index), CryptoHash>,
+    path_item_cache_mapping: HashMap<LeafIndex, Vec<(Level, Index)>>,
+}
+
+impl CachedNodes {
+    fn new() -> Self {
+        Self {
+            inner: HashMap::new(),
+            path_item_cache_mapping: HashMap::new(),
+        }
+    }
+
+    fn extend_from_given(&mut self, given_nodes: &[NodeCoordinates], leaf_index: LeafIndex) {
+        if given_nodes.len() == 0 {
+            return;
+        }
+
+        given_nodes.iter().for_each(|node| {
+            let NodeCoordinates { index, level, hash } = node;
+            if self.inner.get(&(*level, *index)).is_some() {
+                return;
+            }
+            self.inner.insert((*level, *index), hash.unwrap());
+            let e = self
+                .path_item_cache_mapping
+                .entry(leaf_index)
+                .or_insert_with(Vec::new);
+            e.push((*level, *index));
+        });
+    }
+}
+
+impl<HF: HostFunctions> ProofBatchVerifier<HF> {
+    pub fn new() -> Self {
+        Self {
+            cached_nodes: CachedNodes::new(),
+            _hf: PhantomData::default(),
+        }
+    }
+
+    /// Computes the root hash of a given merkle proof and item hash
+    /// It will update the cache of intermediate nodes so that they do not have
+    /// to be recomputed
+    pub fn calculate_root_hash(&mut self, proof: &MerklePath, item_hash: CryptoHash) -> CryptoHash {
+        // trivial example, where proof is empty
+        if proof.len() == 0 {
+            return CryptoHash::default();
+        }
+
+        // the first element is somewhat different, since the caller is passing the item's hash
+        let (_, node_coordinates_to_calculate) = self.get_node_coordinates(proof);
+        let nodes_to_calculate = node_coordinates_to_calculate.len();
+
+        let sibling_item = &proof[0];
+
+        // calculate the hash for the leaf level by hashing the item_hash given and its sibling (provided in the proof)
+        let hash = match sibling_item.direction {
+            Direction::Left => hash_borsh::<_, HF>(&(sibling_item.hash, item_hash)),
+            Direction::Right => hash_borsh::<_, HF>(&(item_hash, sibling_item.hash)),
+        };
+
+        let NodeCoordinates { index, level, .. } =
+            &node_coordinates_to_calculate[nodes_to_calculate - 0 - 1];
+        let cached_value = self.cached_nodes.inner.get(&(*level, *index));
+
+        match cached_value {
+            None => {
+                self.cached_nodes.inner.insert((*level, *index), hash);
+            }
+            Some(parent_hash) => {
+                // ensure that, if the value was cached it matches the calculation made above
+                // this is important, otherwise when most of the intermediates nodes are cached, if this check
+                // is not made, a wrong proof could be passed and stil "yield" the right root hash
+                assert_eq!(parent_hash, &hash);
```

Comment:
Don't we want to reflect it in the return type? (using Option or Result) Because user may submit an invalid proof and trigger panic

Reply:
I believe you're right.

Reply:
I did it in the context of a trusted setup (our own relayer). But it will run on the light client. So it's best to flag it differently